doc/man/man1/strlog.1.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man1/strlog.1.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH STRLOG 1 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ User Commands"
.\"
.\"
.SH NAME
.B strlog
\- write a message to the \fISTREAMS\fR logger
.\"
.\"
.SH SYNOPSIS
.ad l
.HP \w'strlog\ 'u
\fBstrlog\fR
[\fIOPTIONS\fR]
.\"[\fB-ids\fR]
[\fB-s\fR]
[\fB-M\fR\ \fImid\fR]
[\fB-S\fR\ \fIsid\fR]
[\fB-p\fR\ \fIpriority\fR]
[\fB-f\fR\ \fIfile\fR]
.\"[\fB-t\fR\ \fItag\fR]
.\"[\fB-u\fR\ \fIsocket\fR]
[\fImessage\fR\ \fI...\fR]
.PD 0
.HP
\fBstrlog\fR {{\fB-h\fR|\fB--help\fR}|{\fB-V\fR|\fB--version\fR}|{\fB-C\fR|\fB--copying\fR}}
.PD
.ad b
.\"
.\"
.SH DESCRIPTION
.PP
.B strlog
is a shell script interface to the
.IR STREAMS -based
logger,
.BR strlog (4).
.B strlog
accepts messages on the command line
.RI ( message )
or from
.I file
using the
.B -f
option, and delivers those messages to the
.I STREAMS
logger,
.BR strlog (4).
The
.BR -M " and " -S
options can be used to set the
.I STREAMS
module identifier,
.IR mid ,
and unit identifier,
.IR sid .
These identifiers, once set, are applicable to all messages logged.
The
.B -p
option can be used to set the logging
.I facility
and
.I level
as described below.
The priority, once set, is applicable to all messages logged.
.PP
Normally,
.B strlog
generates a message and passes it to the
.I STREAMS
logger,
using the
.BR putmsg (2s)
interface to the log driver,
.BR strlog (4).
Using the
.B -s
option, the user can request that the message be formatted and logged to standard error,
.IR stderr ,
as well as the
.I STREAMS
logger.
.PP
The
.BR -t ", " -i ", " -u " and " -d
options are provided for compatibility with
.BR logger (1),
and are ignored.
.\"
.\"
.SH OPTIONS
.PP
.TP
\fB-M\fR, \fB--mid\fR \fImid\fR
Specifies the module identifier,
.IR mid ,
to be included in the
.I STREAMS
log.  If this option is not present, the
.I mid
defaults to zero
.RB ( 0 ).
The value of
.IR mid ,
specified last with this option (or by default) applies to all messages specified on the command
line, contained in
.IR file ,
or read from
.IR stdin .
.TP
\fB-S\fR, \fB--sid\fR \fIsid\fR
Specifies the unit identifier,
.IR sid ,
to be included in the
.I STREAMS
log.  If this option is not present, the
.I sid
defaults to zero
.RB ( 0 ).
The value of
.IR sid ,
specified last with this option (or by default) applies to all messages specified on the command
line, contained in
.IR file ,
or read from
.IR stdin .
.TP
\fB-p\fR, \fB--priority\fR \fIpriority\fR
Specify the concatenated system log facility and level for the log message.
.I priority
can either be specified as a numeric value, or as a symbolic name using a dot notation:
.IR facility . level .
If this option is not specified,
.I priority
defaults to
.BR user . info .
The value of
.IR priority ,
specified last with this option (or by default) applies to all messages specified on the command
line, contained in
.IR file ,
or read from
.IR stdin .
.TP
\fB-f\fR, \fB--file\fR \fIfile\fR
Provided for compatibility with
.BR logger (1);
when this option is specified, any
.I message
arguments are ingored and logger input is taken from
.IR file .
If
.I file
is
.RB \(lq - \(rq,
then
.I stdin
is read instead of a file.
Each line of input read from
.I file
is treated as a separate
.I message
and a separate invocation of
.BR strlog .
That is, each line of the input is formatted and passed to the
.I STREAMS
logger in the same fashion as though each line of input were specified as the
.I message
string argument to a separate
.B strlog
command.
The
.IR mid ", " sid " and " priority
specified on the command line (or arrived at by default) are applied to each line of the input.
.TP
\fB-s\fR, \fB--stderr\fR
Provided for compatibility with
.BR logger (1);
this option causes the output to also be written to
.IR stderr ,
in a format similar to that provided by the
.I STREAMS
.BR strace (8)
and
.BR strerr (8)
loggers.
See
.BR strace (8)
and
.BR strerr (8)
for a description of those formats.
.TP
\fB-t\fR, \fB--tag\fR \fItag\fR
Provided for compatibility with
.BR logger (1);
this option is ignored.
.TP
\fB-i\fR, \fB--id\fR
Provided for compatibility with
.BR logger (1);
this option is ignored.
.TP
\fB-u\fR, \fB--socket\fR \fIdevice\fR
Provided for compatibility with
.BR logger (1);
this option is ignored.
.TP
\fB-d\fR, \fB--datagram\fR
Provided for compatibility with
.BR logger (1);
this option is ignored.
.TP
\fB-D\fR, \fB--debug\fR [\fIlevel\fR]
Increases or sets to \fIlevel\fR the debug verbosity, typically for debugging purposes.
.TP
\fB-v\fR, \fB--verbose\fR [\fIlevel\fR]
Increases or sets to \fIlevel\fR the output verbosity, typically for diagnostic purposes.
.TP
\fB-q\fR, \fB--quiet\fR
Suppresses normal output.
(Equivalent to
.BR --verbose = 0 .)
.TP
\fB-h\fR, \fB--help\fR, \fB-?\fR, \fB--?\fR
When this option is encountered,
.B strlog
prints usage information and then exits.  No additional option scanning or other actions are
performed.
.TP
\fB-V\fR, \fB--version\fR
When this option is encountered,
.B strlog
prints version information and then exits.  No additional option scanning or other actions are
performed.
.TP
\fB-C\fR, \fB--copying\fR
When this option is encountered,
.B strlog
prints version information and then exits.  No additional option scanning or other actions are
performed.
.\"
.\"
.SH ARGUMENTS
.PP
.TP
\fR[\fImessage\fR \fI...\fR]
Zero or more message strings to be logged, concatentated together with a single space between them.
If no
.I message
argument is specified, and option
.B -f

NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.


src/util/strlog.c
/*****************************************************************************

 @(#) File: src/util/strlog.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/util/strlog.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 *  Linux Fast-STREAMS Utility: strlog - logs STREAMS error or trace messages
 */

#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE 600
#endif

#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <ctype.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/ioctl.h>

#ifdef _GNU_SOURCE
#include <getopt.h>
#endif

#include <time.h>
#include <sys/time.h>
#include <stropts.h>
#include <sys/strlog.h>
#define SYSLOG_NAMES
#include <syslog.h>
#include <sys/sad.h>

static int debug = 0;
static int output = 1;

static void
copying(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	(void) fprintf(stdout, "\
--------------------------------------------------------------------------------\n\
%1$s\n\
--------------------------------------------------------------------------------\n\
Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>\n\
Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>\n\
Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>\n\
\n\
All Rights Reserved.\n\
--------------------------------------------------------------------------------\n\
This program is free software: you can  redistribute it  and/or modify  it under\n\
the terms of the  GNU Affero  General  Public  License  as published by the Free\n\
Software Foundation, version 3 of the license.\n\
\n\
This program is distributed in the hope that it will  be useful, but WITHOUT ANY\n\
WARRANTY; without even  the implied warranty of MERCHANTABILITY or FITNESS FOR A\n\
PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.\n\
\n\
You should have received a copy of the  GNU Affero General Public License  along\n\
with this program.   If not, see <http://www.gnu.org/licenses/>, or write to the\n\
Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n\
--------------------------------------------------------------------------------\n\
U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on behalf\n\
of the U.S. Government (\"Government\"), the following provisions apply to you. If\n\
the Software is supplied by the Department of Defense (\"DoD\"), it is classified\n\
as \"Commercial  Computer  Software\"  under  paragraph  252.227-7014  of the  DoD\n\
Supplement  to the  Federal Acquisition Regulations  (\"DFARS\") (or any successor\n\
regulations) and the  Government  is acquiring  only the  license rights granted\n\
herein (the license rights customarily provided to non-Government users). If the\n\
Software is supplied to any unit or agency of the Government  other than DoD, it\n\
is  classified as  \"Restricted Computer Software\" and the Government's rights in\n\
the Software  are defined  in  paragraph 52.227-19  of the  Federal  Acquisition\n\
Regulations (\"FAR\")  (or any successor regulations) or, in the cases of NASA, in\n\
paragraph  18.52.227-86 of  the  NASA  Supplement  to the FAR (or any  successor\n\
regulations).\n\
--------------------------------------------------------------------------------\n\
Commercial  licensing  and  support of this  software is  available from OpenSS7\n\
Corporation at a fee.  See http://www.openss7.com/\n\
--------------------------------------------------------------------------------\n\
", ident);
}

static void
version(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	(void) fprintf(stdout, "\
%1$s (OpenSS7 %2$s) %3$s (%4$s)\n\
Written by Brian Bidulock.\n\
\n\
Copyright (c) 2008, 2009, 2010, 2011, 2015  Monavacon Limited.\n\
Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008  OpenSS7 Corporation.\n\
Copyright (c) 1997, 1998, 1999, 2000, 2001  Brian F. G. Bidulock.\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\
\n\
Distributed by OpenSS7 under GNU Affero General Public License Version 3,\n\
with conditions, incorporated herein by reference.\n\
\n\
See `%1$s --copying' for copying permissions.\n\
", NAME, PACKAGE, VERSION, PACKAGE_ENVR " " PACKAGE_DATE);
}

static void
usage(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	(void) fprintf(stderr, "\
Usage:\n\
    %1$s [options] [MESSAGE ...]\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
", argv[0]);
}

static void
help(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	(void) fprintf(stdout, "\
Usage:\n\
    %1$s [options] [MESSAGE ...]\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
Arguments:\n\
    MESSAGE\n\
        a message string to be sent to the logger\n\
        if no message string, and no -f option, read from stdin\n\
Options:\n\
    -M, --mid MID\n\
        module id for STREAMS logger\n\
    -S, --sid SID\n\
        unit id for STREAMS logger\n\
    -p, --priority PRIORITY\n\
        number or \"facility.level\" notation PRIORITY\n\
    -E, --error\n\
        generate error log (the default if nothing specified)\n\
    -T, --trace\n\
        generate trace log\n\
    -N, --notify\n\
        notify the system administrator (by mail)\n\
    -c, --console\n\
        display log to the console\n\
    -f, --file FILE\n\
        read messages to log from FILE\n\
    -s, --stderr\n\
        also log messages to standard output\n\
    -t, --tag TAG\n\
        for compatibility with logger(1); ignored\n\
    -i, --id\n\
        for compatibility with logger(1); ignored\n\
    -u, --socket DEVICE\n\
        for compatibility with logger(1); ignored\n\
    -d, --datagram\n\
        for compatibility with logger(1); ignored\n\
    -q, --quiet\n\
        suppress normal output (equivalent to --verbose=0)\n\
    -D, --debug [LEVEL]\n\
        increment or set debug LEVEL [default: 0]\n\
    -v, --verbose [LEVEL]\n\
        increment or set output verbosity LEVEL [default: 1]\n\
        this option may be repeated.\n\
    -h, --help, -?, --?\n\
        print this usage information and exit\n\
    -V, --version\n\
        print version and exit\n\
    -C, --copying\n\
        print copying permission and exit\n\
", argv[0]);
}

void
log_line(short mid, short sid, char priority, int logflags, unsigned long seq, const char *line,
	 int stderr_also)
{
	if (stderr_also) {
		const char *flags = "???";
		struct timespec tv = { 0, 0 };

#ifdef CLOCK_MONOTONIC
		clock_gettime(CLOCK_MONOTONIC, &tv);
#endif
		switch (logflags & (SL_TRACE | SL_FATAL | SL_NOTIFY)) {
		case 0:
			flags = "   ";
			break;
		case SL_NOTIFY:
			flags = "  N";
			break;
		case SL_FATAL:
			flags = " F ";
			break;
		case SL_FATAL | SL_NOTIFY:
			flags = " FN";
			break;
		case SL_TRACE:
			flags = "T  ";
			break;
		case SL_TRACE | SL_NOTIFY:
			flags = "T N";
			break;
		case SL_TRACE | SL_FATAL:
			flags = "TF ";
			break;
		case SL_TRACE | SL_FATAL | SL_NOTIFY:
			flags = "TFN";
			break;
		}
		fprintf(stderr, "%3lu %12lu %12lu %s %5hd %5hd %s\n", seq, time(NULL), tv.tv_sec,
			flags, mid, sid, line);
	}
}

int
main(int argc, char *argv[])
{
	int options;
	int my_logflags = 0;		/* user specified flags */
	int xflags = 0;			/* extra flags */
	int stderr_also = 0;		/* log also to stderr */
	int logflags = SL_ERROR;	/* logger flags */
	int priority = LOG_USER | LOG_DEBUG;	/* facility and level */
	short mid = 0;			/* module identifier */
	short sid = 0;			/* subunit identifier */
	char filename[256] = "";	/* file name */
	int use_file = 0;		/* use file in filename */
	int read_stdin = 0;		/* read from stdin */

	for (options = 0;; options++) {
		int c, val;

#if defined _GNU_SOURCE
		int option_index = 0;
		/* *INDENT-OFF* */
		static struct option long_options[] = {
			{"mid",		required_argument,	NULL, 'M'},
			{"sid",		required_argument,	NULL, 'S'},
			{"priority",	required_argument,	NULL, 'p'},
			{"error",	no_argument,		NULL, 'E'},
			{"trace",	no_argument,		NULL, 'T'},
			{"notify",	no_argument,		NULL, 'N'},
			{"console",	no_argument,		NULL, 'c'},
			{"file",	required_argument,	NULL, 'f'},
			{"stderr",	no_argument,		NULL, 's'},
			{"tag",		required_argument,	NULL, 't'},
			{"id",		no_argument,		NULL, 'i'},
			{"socket",	required_argument,	NULL, 'u'},
			{"datagram",	no_argument,		NULL, 'd'},
			{"quiet",	no_argument,		NULL, 'q'},
			{"debug",	optional_argument,	NULL, 'D'},
			{"verbose",	optional_argument,	NULL, 'v'},
			{"help",	no_argument,		NULL, 'h'},
			{"version",	no_argument,		NULL, 'V'},
			{"copying",	no_argument,		NULL, 'C'},
			{"?",		no_argument,		NULL, 'H'},
			{ 0, }
		};
		/* *INDENT-ON* */

		c = getopt_long_only(argc, argv, "M:S:p:ETNcf:st:iu:dqD::v::hVC?", long_options,
				     &option_index);
#else
		c = getopt(argc, argv, "M:S:p:ETNcf:st:iu:dqD::v::hVC?");
#endif
		if (c == -1) {
			if (debug)
				fprintf(stderr, "%s: done options processing\n", argv[0]);
			break;
		}
		switch (c) {
		case 'M':	/* -M, --mid MID */
			if ((val = strtol(optarg, NULL, 0)) < 0 || val > 0xffffUL)
				goto bad_option;
			mid = val;
			continue;
		case 'S':	/* -S, --sid SID */
			if ((val = strtol(optarg, NULL, 0)) < 0 || val > 0xffffUL)
				goto bad_option;
			sid = val;
			continue;
		case 'p':	/* -p, --priority PRIORITTY */
		{
			CODE *fac = NULL;
			CODE *pri = NULL;

			if (isdigit(*optarg)) {
				/* its a number, use it */
				if ((val = strtol(optarg, NULL, 0)) < 0)
					goto bad_option;
				for (fac = facilitynames; fac->c_name; fac++)
					if (LOG_FAC(val) == LOG_FAC(fac->c_val))
						break;
				if (!fac->c_name)
					goto bad_option;
				for (pri = prioritynames; pri->c_name; pri++)
					if (val == pri->c_val)
						break;
				if (!pri->c_name)
					goto bad_option;
				priority = val;
			} else {
				char facil_name[9];
				char level_name[9];

				if (sscanf(optarg, "%8[a-z].%8[a-z]", facil_name, level_name) != 2)
					goto bad_option;
				if (facil_name[0] == '\0')
					goto bad_option;
				for (fac = facilitynames; fac->c_name; fac++)
					if (strncmp(fac->c_name, facil_name, 9) == 0)
						break;
				if (!fac->c_name)
					goto bad_option;
				if (level_name[0] == '\0')
					goto bad_option;
				for (pri = prioritynames; pri->c_name; pri++)
					if (strncmp(pri->c_name, level_name, 9) == 0)
						break;
				if (!pri->c_name)
					goto bad_option;
			}
			/* things are ok now */
			priority = LOG_MAKEPRI(LOG_FAC(fac->c_val), pri->c_val);
			switch (LOG_PRI(priority)) {
			case LOG_EMERG:
				xflags |= SL_FATAL;
				break;
			case LOG_ALERT:
			case LOG_CRIT:
			case LOG_ERR:
				break;
			case LOG_WARNING:
				xflags |= SL_WARN;
				break;
			case LOG_NOTICE:
				xflags |= SL_NOTE;
				break;
			case LOG_INFO:
			case LOG_DEBUG:
				break;
			}
			continue;
		}
		case 'E':	/* -E, --error */
			my_logflags |= SL_ERROR;
			continue;
		case 'T':	/* -T, --trace */
			my_logflags |= SL_TRACE;
			continue;
		case 'N':	/* -N, --notify */
			my_logflags |= SL_NOTIFY | SL_ERROR;
			continue;
		case 'c':	/* -c, --console */
			my_logflags |= SL_CONSOLE | SL_ERROR;
			continue;
		case 'f':	/* -f, --file FILE */
			if (strnlen(optarg, sizeof(filename) + 1) > sizeof(filename))
				goto bad_option;
			strncpy(filename, optarg, sizeof(filename));
			if (filename[0] == '\0')
				goto bad_option;
			if (strncmp(filename, "-", sizeof(filename)) == 0) {
				filename[0] = '\0';
				read_stdin = 1;
			} else
				use_file = 1;
			continue;
		case 's':	/* -s, --stderr */
			stderr_also = 1;
			continue;
		case 't':	/* -t, --tag TAG */
		case 'i':	/* -i, --id */
		case 'u':	/* -u, --socket DEVICE */
		case 'd':	/* -d, --datagram */
			/* these four are ignored for compatibiltiy with logger(1) */
			continue;
		case 'q':	/* -q, --quiet */
			if (debug)
				fprintf(stderr, "%s: suppressing normal output\n", argv[0]);
			output = 0;
			debug = 0;
			continue;
		case 'D':	/* -D, --debug */
			if (debug)
				fprintf(stderr, "%s: increasing debug verbosity\n", argv[0]);
			if (optarg == NULL) {
				debug++;
				continue;
			}
			if ((val = strtol(optarg, NULL, 0)) < 0)
				goto bad_option;
			debug = val;
			continue;
		case 'v':	/* -v, --verbose */
			if (debug)
				fprintf(stderr, "%s: increasing output verbosity\n", argv[0]);
			if (optarg == NULL) {
				output++;
				continue;
			}
			if ((val = strtol(optarg, NULL, 0)) < 0)
				goto bad_option;
			output = val;
			continue;
		case 'h':	/* -h, --help */
		case 'H':	/* -H, --? */
			if (debug)
				fprintf(stderr, "%s: printing help message\n", argv[0]);
			help(argc, argv);
			exit(0);
		case 'V':	/* -V, --version */
			if (debug)
				fprintf(stderr, "%s: printing version message\n", argv[0]);
			version(argc, argv);
			exit(0);
		case 'C':	/* -C, --copying */
			if (debug)
				fprintf(stderr, "%s: printing copying message\n", argv[0]);
			copying(argc, argv);
			exit(0);
		case '?':
		default:
		      bad_option:
			optind--;
			goto bad_nonopt;
		      bad_nonopt:
			if (output > 0 || debug > 0) {
				if (optind < argc) {
					const char *delim = "";

					fprintf(stderr, "%s: syntax error near '", argv[0]);
					while (optind < argc) {
						fprintf(stderr, "%s%s", delim, argv[optind++]);
						delim = " ";
					}
					fprintf(stderr, "'\n");
				} else {
					fprintf(stderr, "%s: missing option or argument", argv[0]);
					fprintf(stderr, "\n");
				}
				fflush(stderr);
				usage(argc, argv);
			}
			exit(2);
		}
	}
	if (optind >= argc) {
		char buffer[1024];
		char *line;
		unsigned long lineno;
		FILE *f = stdin;

		/* no message argument provided */
		if (use_file == 0 && read_stdin == 0)
			/* nowhere to get messages from */
			goto bad_nonopt;
		if (use_file) {
			if ((f = fopen(filename, "r")) == NULL) {
				if (output || debug)
					perror(__FUNCTION__);
				exit(1);
			}
		}
		logflags = (my_logflags ? my_logflags : logflags) | xflags;
		/* read the file one line at a time */
		for (lineno = 1; (line = fgets(buffer, sizeof(buffer), f)) != NULL; lineno++) {
			size_t len;

			line[sizeof(buffer)] = '\0';
			/* find newline */
			len = strnlen(line, sizeof(buffer));
			/* chop the newline */
			if (len > 0 && line[len - 1] == '\n') {
				len--;
				line[len] = '\0';
			}
			log_line(mid, sid, priority, logflags, lineno, line, stderr_also);
		}
		fclose(f);
	} else {
		/* a message argument(s) was provided */
		if (read_stdin)
			goto bad_nonopt;
		logflags = (my_logflags ? my_logflags : logflags) | xflags;
		/* log messages one string at a time */
		for (; optind < argc; optind++)
			log_line(mid, sid, priority, logflags, (argc - optind), argv[optind],
				 stderr_also);
	}
	exit(0);
}
